slsDetectorServers/slsDetectorServer/src/sharedMemory.c

```diff
@@ -0,0 +1,200 @@
+#include "sharedMemory.h"
+#include "clogger.h"
+#include "sls_detector_defs.h"
+
+#include <errno.h> // errno
+#include <fcntl.h> // O_CREAT, O_TRUNC..
+#include <string.h>
+#include <sys/mman.h> // shared memory
+#include <sys/stat.h> // fstat
+#include <unistd.h>
+
+#include <sys/ipc.h>
+#include <sys/shm.h>
+#include <sys/types.h>
+
+#define SHM_NAME    "sls_server_shared_memory"
+#define SHM_VERSION 0x200625
+#define SHM_KEY     5678
+
+typedef struct Memory {
+    int version;
+    sem_t sem;
+    int scanStatus;
+    int scanStop;
+#ifdef VIRTUAL
+    int status;
+    int stop;
+#endif
+} sharedMem;
+
+sharedMem *shm = NULL;
+char shmMess[MAX_STR_LENGTH];
+int shmFd = -1;
+
+extern int isControlServer;
+
+char *sharedMemory_getError() { return shmMess; }
+
+void sharedMemory_print() {
+    LOG(logINFO, ("%s Shared Memory:\n", isControlServer ? "c" : "s"));
+    LOG(logINFO,
+        ("%s version:0x%x\n", isControlServer ? "c" : "s", shm->version));
+#ifdef VIRTUAL
+    LOG(logINFO, ("%s status: %d\n", isControlServer ? "c" : "s", shm->status));
+    LOG(logINFO, ("%s stop: %d\n", isControlServer ? "c" : "s", shm->stop));
+#endif
+}
+
+int sharedMemory_create(int port) {
+    memset(shmMess, 0, MAX_STR_LENGTH);
+
+    // if sham existed, delete old shm and create again
+    shmFd =
+        shmget(SHM_KEY + port, sizeof(sharedMem), IPC_CREAT | IPC_EXCL | 0666);
+    if (shmFd == -1 && errno == EEXIST) {
+        LOG(logWARNING, ("Removing old shared memory\n"));
+        // cuz of unknown previous shm size, it has to  be deleted like this
+        system(
+            "for i in seq `ipcs -m | cut -d ' ' -f1`; do ipcrm -M $i; done;");
+        shmFd = shmget(SHM_KEY + port, sizeof(sharedMem),
+                       IPC_CREAT | IPC_EXCL | 0666);
+    }
+    if (shmFd == -1) {
+        sprintf(shmMess, "Create shared memory failed: %s\n", strerror(errno));
+        LOG(logERROR, (shmMess));
+        return 0;
+    }
+    LOG(logINFO, ("Shared memory created\n"));
+    if (!sharedMemory_attach()) {
+        return 0;
+    }
+    sharedMemory_initialize();
+    return 1;
+}
+
+void sharedMemory_initialize() {
```

Comment:
i was passing a pointer before.. but then for every function, you need to do it.. and since we have just one.. it made sense to abstract it by putting it inside the sharedmemory

Reply:
Ok, then we keep it this way.